### PR TITLE
348 add timeout to fetch requests

### PR DESCRIPTION
**Pull request by rajkambo**
* [x]  Passed [sanity tests](http://bit.ly/sttm-sanity-tests).
* [x]  Ran `npm test` & fixed newly introduced lint errors.
* [x]  Checked console for errors.

<!-- New to markdown? Simply put an 'x' between [ ] to check it. Like [x] this. (No spaces).

### CODE REVIEW

File: src/js/components/Fetch.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 
+const TIMEOUT = 10000; // 10 seconds
```
Comment (bogas04) [inline src/js/components/Fetch.js]: Could you refactor it so that Fetch accepts timeout as a prop and defaults to 10s?

Comment (rajkambo) [reply src/js/components/Fetch.js]: Yup, for sure! 👍

**Comment by rajkambo** [comment]
So I updated the PR with timeout getting passed as a prop, but I found that timing out seems to cause #365:
image

Will do some more digging into how we can fix this 👍

**Comment by bogas04** [comment]
@rajkambo were you able to fix the bug?

**Comment by rajkambo** [comment]
@bogas04 just mulling over what would be the best way to handle the issue. So the root cause of this error and #365 is the same -- a trying to access fields from a the null "data" object here: https://github.com/KhalisFoundation/sttm-web/blob/master/src/js/pages/Search/index.js#L44. 

This is caused by the error being null here: https://github.com/KhalisFoundation/sttm-web/blob/master/src/js/pages/PageLoader.js#L17. 

Since error the error is null here data (null) is passed on to the children and a null exception occurs leading to the generic: https://github.com/KhalisFoundation/sttm-web/blob/master/src/js/Root.js#L29.

In my specific case (on timed out Fetch requests), I could modify the setTimeout call to have a third argument and make it something like "Fetch request timed out", this would then ensure that error isn't null as it is passed on here: https://github.com/KhalisFoundation/sttm-web/pull/473/files#diff-10ef1ed613a66115bbbefad359a37a6cR67.

In this case we end up back at: https://github.com/KhalisFoundation/sttm-web/blob/master/src/js/pages/PageLoader.js#L17. Where an error is thrown when "error" isn't null. And since this error doesn't get caught anywhere we end up right here: https://github.com/KhalisFoundation/sttm-web/blob/master/src/js/Root.js#L29.

And in since the error goes uncaught, we end up in a state where we can't click anything.. 

Going to do some more reading on how exactly error handling works in React, but would be great if you had any suggestions :)

**Comment by rajkambo** [comment]
Update: I think I got the ideal fix! Will commit soon 👍

File: src/js/components/HandleSearchError.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import { ACTIONS, errorEvent } from '../util/analytics';
+import GenericError, { SachKaur } from './GenericError';
+import { TYPES, SOURCES, PLACEHOLDERS, TEXTS } from '../constants';
+
+export default class HandleSearchError extends React.PureComponent {
+  static propTypes = {
+    q: PropTypes.string.isRequired,
+    type: PropTypes.number.isRequired,
+    source: PropTypes.string.isRequired,
+    children: PropTypes.node.isRequired,
```
Comment (rajkambo) [inline src/js/components/HandleSearchError.js]: Wasn't sure about the type of "children" but read this: https://stackoverflow.com/questions/42122522/reactjs-what-should-the-proptypes-be-for-this-props-children and settled on "node".

File: src/js/components/HandleSearchError.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import React from 'react';
```
Comment (rajkambo) [inline src/js/components/HandleSearchError.js]: Created this component for the error handling approach suggested here: https://reactjs.org/blog/2017/07/26/error-handling-in-react-16.html.

**Comment by rajkambo** [comment]
@bogas04 took a while to dig into this issue, and I fixed this specific error case, but it looks like we need to fix error handling as a whole in the project (as the catch all we had in Root.js isn't going to cut it). I left a detailed comment in #365  on why the links aren't working and what I think we need to do to reform our error handling👍

**Comment by rajkambo** [comment]
@bogas04 could I get the thumbs up when you get the time 👍 Would be great to wrap this one up :D

**Comment by bogas04** [comment]
@rajkambo do we need HandleSearchError or GenericError catches the fetch timeout fine?

**Comment by bogas04** [comment]
I added the logic at our single ErrorBoundary (Layout) to keep logic simple.

@rajkambo thank you so much!

screen shot 2018-06-15 at 7 42 04 pm

**Pull request merged**